### PR TITLE
SPOI-4640: Changed reading of portTypeInfo, port annotations in operator discove…

### PR DESCRIPTION
**Pull request by ishark**
…rer to read from ASM

Added test case for operator discoverer
Updated Type Discovery tests to getPortInfo via ASM instead of reflection API

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 1]
```diff
@@ -455,11 +459,10 @@ public void setClassNode(CompactClassNode classNode)
         }
         result.add(node.typeName);
       }
-      
     }
     return result;
   }
-
+
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Fix editor settings.

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 2]
```diff
@@ -479,11 +482,110 @@ public JSONObject describeClass(String clazzName) throws JSONException
...
+        Type t = fieldType;
+        if (fieldType instanceof ParameterizedTypeNode) {
+          // TODO: Right now getPortInfo assumes a single parameterized type
+          t = ((ParameterizedTypeNode) fieldType).getActualTypeArguments()[0];
+        } else {
+          // TODO: Check behavior for Ports not using Default Input/output ports
+          LOG.debug("Field is of type {}", fieldType.getClass());
+        }
+
+        LOG.debug("field {} class {} type {}", port.getName(), fieldType,
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Remove extra logging?

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 3]
```diff
@@ -479,11 +482,110 @@ public JSONObject describeClass(String clazzName) throws JSONException
...
+            t.getClass());
+        JSONObject meta = new JSONObject();
+        try {
+          meta.put("name", port.getName()); 
+          setTypes(meta, t, typeReplacement);
+          portInfo.add(meta);
+        } catch (Exception e) {
+          DTThrowable.rethrow(e);
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Use DTThrowable.wrapIfChecked

File: engine/src/main/java/com/datatorrent/stram/webapp/asm/CompactFieldNode.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+public class CompactFieldNode {
+  private String name;
+  private String desc;
+  private String signature;
+  
+  private List<Object> visibleAnnotations;
+  private FieldSignatureVisitor fieldSignatureNode;
+//  
+//  public CompactFieldNode(FieldNode fn)
+//  {
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/webapp/asm/CompactFieldNode.java]: Cleanup please.

File: engine/src/main/java/com/datatorrent/stram/webapp/asm/FieldNode.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    super(access, name, desc, signature, value);
+  }
+  
+  @Override
+  public void visitEnd()
+  {
+    super.visitEnd();
+    String methodString = signature != null ? signature : desc;
+    // System.out.println(methodString);
+    // System.out.println("RRRRRRRRRRRR" + methodString + "%%%%%" + name);
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/webapp/asm/FieldNode.java]: Cleanup

File: engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java [hunk 6]
```diff
@@ -51,10 +51,10 @@ public StringOutputPort(Operator operator) {
       public void process(T1 tuple) {
       }
     };
-    final OutputPort<T2> outportT2 = new DefaultOutputPort<T2>();
-    final OutputPort<Number> outportNumberParam = new DefaultOutputPort<Number>();
-    final StringOutputPort outportString = new StringOutputPort(this);
-    final OutputPort<List<T0>> outportList = new DefaultOutputPort<List<T0>>();
+    public transient final OutputPort<T2> outportT2 = new DefaultOutputPort<T2>();
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java]: Why is this change necessary?

File: engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java [hunk 7]
```diff
@@ -81,24 +81,29 @@ public void process(String tuple) {
...
+    
     System.out.println(json.toString(2));
-    System.out.println(td.typeArguments);
 
     ObjectMapper mapper = new ObjectMapper();
     //List<?> l = mapper.convertValue(json, List.class);
     JsonNode root = mapper.readTree(json.toString(2));
     String val = root.get(0).path("name").asText();
-    Assert.assertEquals("port name", "inputT1", val);
+    Assert.assertEquals("port name", "outportT2", val);
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java]: Is this deterministic?

File: engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java [hunk 8]
```diff
@@ -140,9 +149,15 @@ public void testTypeDiscovery2() throws Exception
...
-    JSONArray json = td.getPortTypes(SubSubClass.class);
-
+    String[] classFilePath = OperatorDiscoveryTest.getClassFileInClasspath();
+    OperatorDiscoverer od = new OperatorDiscoverer(classFilePath);
+    od.buildTypeGraph();
+    
+    JSONObject Desc = od.describeClassByASM(SubSubClass.class.getName());
+    System.out.println(Desc);
+    JSONArray json = Desc.getJSONArray("portTypeInfo");
+    System.out.println(json);
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java]: Remove System.out and logging from test, replace with assertions.

File: engine/src/main/java/com/datatorrent/stram/webapp/OperatorDiscoverer.java [hunk 9]
```diff
@@ -379,77 +378,44 @@ public JSONObject describeOperator(Class<? extends Operator> clazz) throws Excep
...
-              inputPort.put("optional", false); // input port that is not annotated is default to be not optional
-            }
 
-            inputPorts.put(inputPort);
+      try {
+        for (CompactFieldNode field : inputPortfields) {
+          JSONObject inputPort = setFieldAttributes(clazz, field);
+          if(!inputPort.has("optional"))
+          {
+            inputPort.put("optional", false); // input port that is not
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/webapp/OperatorDiscoverer.java]: Check code formatting settings.

File: engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java (hunk 6 above)
Comment (ishark) [inline engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java]: I have added check for public , transient and final fields to be added as ports in ASMUtil, under the assumption that  normally ports would be public , transient and final. I can keep the check for only final fields. Please suggest.

File: engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java (hunk 7 above)
Comment (ishark) [inline engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java]: Yes, it is deterministic. I had by mistake not added public transient to inputT1 operator, so it did not get picked up as port. I have updated it in the next commit.

File: engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java (hunk 6 above)
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java]: Why to build these assumptions into the discovery code? What was the problem with the existing behavior?

Comment (ishark) [inline engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java]: Removed this check in next commit.

File: engine/src/main/java/com/datatorrent/stram/webapp/OperatorDiscoverer.java [hunk 10]
```diff
@@ -510,6 +470,36 @@ else if (clazz.getName().startsWith("com.datatorrent.lib.") ||
...
+        String fieldDesc = oci.fields.get(field.getName());
+        if (fieldDesc != null) {
+          port.put("description", fieldDesc);
+          break;
+        }
+      }
+    }
+
+    List<CompactAnnotationNode> annotations = field.getVisibleAnnotations();
+    CompactAnnotationNode firstAnnotation;
```
Comment (siyuanh) [inline engine/src/main/java/com/datatorrent/stram/webapp/OperatorDiscoverer.java]: We can keep this in compact field node to save space and code

Comment Cancel

File: engine/src/main/java/com/datatorrent/stram/webapp/asm/ASMUtil.java [hunk 11]
```diff
@@ -134,5 +149,14 @@ private static boolean isPublic(int opCode)
   {
     return (opCode & Opcodes.ACC_PUBLIC) == Opcodes.ACC_PUBLIC;
   }
+
+  private static boolean isTransient(int opCode)
```
Comment (siyuanh) [inline engine/src/main/java/com/datatorrent/stram/webapp/asm/ASMUtil.java]: make util function public

File: engine/src/main/java/com/datatorrent/stram/webapp/asm/CompactAnnotationNode.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datatorrent.stram.webapp.asm;
+
+import java.util.Map;
+
+public class CompactAnnotationNode {
```
Comment (siyuanh) [inline engine/src/main/java/com/datatorrent/stram/webapp/asm/CompactAnnotationNode.java]: Keep it in CompactFieldNode

File: engine/src/main/java/com/datatorrent/stram/webapp/OperatorDiscoverer.java [hunk 13]
```diff
@@ -260,16 +266,23 @@ public void buildTypeGraph()
...
         }
-      } catch (IOException ex) {
-        LOG.warn("Cannot process file {}", f, ex);
       }
-    }
 
+      typeGraph.updatePortTypeInfoInTypeGraph(openJarFiles, openClassFiles);
+    }
+   finally {
+      for (Entry<String, JarFile> entry : openJarFiles.entrySet()) {
```
Comment (siyuanh) [inline engine/src/main/java/com/datatorrent/stram/webapp/OperatorDiscoverer.java]: close open class file as well

File: engine/src/main/java/com/datatorrent/stram/webapp/asm/CompactFieldNode.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datatorrent.stram.webapp.asm;
+
+import java.util.LinkedList;
```
Comment (siyuanh) [inline engine/src/main/java/com/datatorrent/stram/webapp/asm/CompactFieldNode.java]: Remove unused imports

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 15]
```diff
@@ -220,6 +225,71 @@ public void addNode(JarEntry jarEntry, JarFile jar) throws IOException
...
+    TypeGraphVertex tgv = typeGraph.get(Operator.class.getName());
+    Set<TypeGraphVertex> allOperators = new HashSet<TypeGraph.TypeGraphVertex>();
+    getAllDescendents(tgv, allOperators);
+
+    for (TypeGraphVertex operator : allOperators) {
+      try {
+        String path = operator.getJarName();
+        JarFile jar = openJarFiles.get(path);
+        if (jar != null) {
+          java.util.Enumeration<JarEntry> entriesEnum = jar.entries();
```
Comment (siyuanh) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Should only rescan the operator file. Use getEntry(String name) to locate the operator class

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 16]
```diff
@@ -220,6 +225,71 @@ public void addNode(JarEntry jarEntry, JarFile jar) throws IOException
     addNode(jar.getInputStream(jarEntry), jar.getName());
   }
 
+  public void updatePortTypeInfoInTypeGraph(Map<String, JarFile> openJarFiles,
+      Map<String, File> openClassFiles) {
+    TypeGraphVertex tgv = typeGraph.get(Operator.class.getName());
+    Set<TypeGraphVertex> allOperators = new HashSet<TypeGraph.TypeGraphVertex>();
+    getAllDescendents(tgv, allOperators);
```
Comment (siyuanh) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Why not call updatePortInfo recursively?

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 17]
```diff
@@ -220,6 +225,71 @@ public void addNode(JarEntry jarEntry, JarFile jar) throws IOException
...
+      classN.setClassSignatureVisitor(tgv.getClassNode().getCsv());
+      classN.setVisitFields(true);
+      reader.accept(classN, ClassReader.SKIP_CODE);
+      CompactClassNode ccn = tgv.getClassNode();
+      CompactUtil.updateCompactClassPortInfo(classN, ccn);
+      List<CompactFieldNode> prunedFields = new LinkedList<CompactFieldNode>();
+      TypeGraphVertex portVertex = typeGraph.get(Operator.Port.class.getName());
+      for (CompactFieldNode field : ccn.getPorts()) {
+        TypeGraphVertex fieldVertex = typeGraph.get(field.getDescription());
+        if (isAncestor(portVertex, fieldVertex)) {
```
Comment (siyuanh) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Why not filter the fields in ClassNodeType when call the visitFieldMethod

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 18]
```diff
@@ -479,23 +552,147 @@ public JSONObject describeClass(String clazzName) throws JSONException
...
+    for (CompactFieldNode port : ports) {
+        Type fieldType = port.getFieldSignatureNode().getFieldType();
+        Type t = fieldType;
+        if (fieldType instanceof ParameterizedTypeNode) {
+          // TODO: Right now getPortInfo assumes a single parameterized type
+          t = ((ParameterizedTypeNode) fieldType).getActualTypeArguments()[0];
+        } else {
+          // TODO: Check behavior for Ports not using Default Input/output ports
+          TypeGraphVertex portVertex = typeGraph.get(port.getDescription());
+          t = findTypeArgument(portVertex, typeReplacement);
```
Comment (siyuanh) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: We don't need this, the actual type will be replace in setTypes(line 586) method

**Comment by siyuanh** [comment]
I'm OK with all the changes now, if test passes and also verified with gateway. I think we can merge it. There are still some cleanup work to do(especially the describe operator method is still a hybrid of reflection and ASM). But that could be logged as another ticket

File: engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java [hunk 19]
```diff
@@ -129,20 +171,27 @@ public void testTypeDiscovery2() throws Exception
...
   {
-    TypeDiscoverer td = new TypeDiscoverer();
-    JSONArray json = td.getPortTypes(SubSubClass.class);
-
+    String[] classFilePath = OperatorDiscoveryTest.getClassFileInClasspath();
+    OperatorDiscoverer od = new OperatorDiscoverer(classFilePath);
+    od.buildTypeGraph();
+    
+    JSONObject Desc = od.describeClassByASM(SubSubClass.class.getName());
+    System.out.println(Desc);
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java]: Remove and cover with assertions.

File: engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java (hunk 8 above)
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java]: Remove and cover with assertions.

File: engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java [hunk 20]
```diff
@@ -155,7 +204,7 @@ public void testTypeDiscoveryMultiLevel() throws Exception
     Assert.assertEquals("map key type", "java.lang.String", val);
     val = root.get(0).path("typeArgs").get(1).path("type").asText();
     Assert.assertEquals("map value type", "java.lang.Object", val);
-
+
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/webapp/TypeDiscoveryTest.java]: IDE settings - You can tell Eclipse to automatically remove trailing white space on save.
